### DISCUSSION TITLE
How to disable span from 3rd part crate

### DISCUSSION THREAD

**Discussion by gftea** [tokio-rs/tracing#2997] [Q&A]
Hi, I got a lot of span emitted from 3rd party crate like below without module name, and I would like to disable it.

> Logs//node_3/supra_node_logs/supra.log:[2024-06-04T16:13:36.092377Z+00:00] DEBUG new_established_connection{remote_addr=/ip4/127.0.0.1/tcp/52959 id=2 peer=12D3KooWSePvZJ52R1rHpuaLC7eeJQeG532ZQGTaUZhbQ4koAMWa}:Connection::poll: libp2p_swarm::connection: start + duration cannot be presented, halving duration start=Instant { tv_sec: 641149, tv_nsec: 328546041 } duration=9223372036854775807.5s
Logs//node_3/supra_node_logs/supra.log:[2024-06-04T16:13:36.092381Z+00:00] DEBUG new_established_connection{remote_addr=/ip4/127.0.0.1/tcp/52959 id=2 peer=12D3KooWSePvZJ52R1rHpuaLC7eeJQeG532ZQGTaUZhbQ4koAMWa}:Connection::poll: yamux::connection::rtt: sending ping 3758414513    

I setup tracing below, I wonder how to avoid such span traces emitted? Or anything incorrect in below config? Or it is target filter bug?
```rust
    let env_filter = env::var("RUST_LOG")
        .unwrap_or_else(|_| "info".to_string())
        .parse::<Targets>()?;

    let (target_filtered_layer, target_reload_handle) = reload::Layer::new(env_filter);

    tracing_subscriber::registry()
        .with(
            tracing_subscriber::fmt::layer()
                .with_ansi(false)
                .with_timer(timer.clone())
                .with_writer(non_blocking_file_rotater.and(make_writer)),
        )
        .with(target_filtered_layer)
        .try_init()?;

**Comment by kaffarell** [comment]
Hi!
you can use your existing RUST_LOG setup and set a env filter e.g.: `RUST_LOG='none, your_crate=debug'` or `RUST_LOG='third_party_crate=error,your_crate=debug'`
More info on the EnvFilter syntax: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/struct.EnvFilter.html
For more custom filters, you can also create a filter_fn: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/filter/fn.filter_fn.html

**Reply by gftea** [reply]
As you can see from the output span, there is no crate path for the log. I want to disable that specific crate, because I have about 100 crates in my workspace, while I have about 1000 dependencies, it is not effecient to cherry pick each of my crate in a RUST_LOG filtering

**Reply by mladedav** [reply]
I think the output contains the target, it's `yamux::connection`, so you want to disable `yamux`.

**Reply by gftea** [reply]
- I am a bit unclear why its module name is not libp2p-yamux, if I look at the package name in https://github.com/libp2p/rust-libp2p/blob/master/muxers/yamux/Cargo.toml, it should start  `libp2p_yamux` or I misunderstand it
- 2nd, because the same span can be from different modules with same name `new_established_connection`, and I added below filter `[new_established_connection]=off` to try to disable all targets with this name, but it does not work

**Reply by mladedav** [reply]
The crate you linked itself depends on one named [yamux](https://crates.io/crates/yamux). I'd assume that one is the one that actually produces the logs?

Also the target does not need to be the module path, each call may override it to anything else.

As for why the directive you sent does not work, I have no idea, this should be supported but there were some issues with env-filter parsing, so I'll try to check.

And to be clear, setting `yamux=off` solved the noisy events or are they still happening?

**Reply by gftea** [reply]
I updated the traces output examples, just show there are two targets.
 I tried below
1. `yamux=off` :  OK, new_established_connection span is not emitted by yamux
1. `libp2p_swarm::connection=off` : OK, but all new_established_connection are off including yamux.
2. `libp2p_swarm::connection[new_established_connection]=off` : NOK, all span emitted.


The rust-libp2p configure the span like below, see https://github.com/libp2p/rust-libp2p/blob/master/swarm/src/connection/pool.rs#L534
```
        let span = tracing::debug_span!(parent: tracing::Span::none(), "new_established_connection", remote_addr = %endpoint.get_remote_address(), %id, peer = %obtained_peer_id);
        span.follows_from(tracing::Span::current());

        self.executor.spawn(
            task::new_for_established_connection(
                id,
                obtained_peer_id,
                connection,
                command_receiver,
                event_sender,
            )
            .instrument(span),
        )
```
